### PR TITLE
Bulk shard migrator

### PR DESCRIPTION
**Pull request by qelper**
Mostly porting franklin's bulk shard migrator logic to misk

### CODE REVIEW

File: misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+    // try to keep it short with small batches.
+    val idsToDelete = transacter.transaction{ session ->
+      val sourceRecords = loadSourceRecords(session)
+      if (sourceRecords.isEmpty()) {
+        setOf()
+      } else {
+        val existingIds = session.hibernateSession.createNativeQuery(
+            "SELECT id FROM $tableName WHERE $rootColumnName = :target AND id IN (:ids)"
+        ).setParameter("target", targetRoot!!.id)
+        .setParameterList("ids", sourceRecords.keys)
```
Comment (shellderp) [inline misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt]: should be indented more

File: misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+  private val childClass: KClass<C>
+) {
+
+  private val persistenceMetadata: PersistenceMetadata
+  private val keyspace: Keyspace
+
+  private val mutations = ArrayList<Mutation>()
+  private var where: String? = null
+  private var parameters: List<*>? = null
+  private var rootColumnName = "customer_id"
```
Comment (shellderp) [inline misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt]: likely won't work for all services

File: misk-hibernate/src/test/kotlin/misk/hibernate/ShardIdGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package misk.hibernate
+
+class ShardIdGenerator constructor(private val shard: Shard) {
```
Comment (shellderp) [inline misk-hibernate/src/test/kotlin/misk/hibernate/ShardIdGenerator.kt]: where is this used?

File: misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+import java.util.ArrayList
+import java.util.HashMap
+import java.util.stream.Collectors
+import java.util.stream.Collectors.joining
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.reflect.KClass
+import kotlin.reflect.full.isSubclassOf
+
+class BulkShardMigrator<R : DbRoot<R>, C : DbChild<R, C>> private constructor(
```
Comment (mightyguava) [inline misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt]: docs please! What's this do? How does it relate to sql migrations? What is the API for this class?

File: misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt [hunk 5]
```diff
@@ -0,0 +1,435 @@
+package misk.hibernate
```
Comment (mightyguava) [inline misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt]: I'm not sure if this should be in the hibernate package. It's a specific workflow for applications to migrate data. Maybe misk.hibernate.migrate

File: misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt (hunk 2 above)
Comment (qelper) [reply misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt]: yeah i have to do to look it up based on annotations

File: misk-hibernate/src/test/kotlin/misk/hibernate/ShardIdGenerator.kt (hunk 3 above)
Comment (qelper) [reply misk-hibernate/src/test/kotlin/misk/hibernate/ShardIdGenerator.kt]: no where, was using it in testing. Removed it.

File: misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt (hunk 4 above)
Comment (qelper) [reply misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt]: sure, added some docs

File: misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt (hunk 5 above)
Comment (qelper) [reply misk-hibernate/src/main/kotlin/misk/hibernate/BulkShardMigrator.kt]: moved to `misk.hibernate.migrate`

**Pull request merged**